#!/bin/bash
##
## Evaluates a fusion calling procedure on simulated data with ground truths
##
## Started 11th Nov 2024
##
## Izaskun Mallona

NTHREADS=30

:<<EOF
Expected reads follow this pattern, being the `/` the splicing site or fusion site
  apparently only 16 nt on each side of the fusion are expected (Giulia dixit, something about
  rois not being efficient)
The actual content is available as ./data/reference_fusions.fa

ENST00000372348.7						GTTTTTGTGGAACATG/AAGCCCTTCAGCGGCC # Human ABL1 across fusion (no fusion)
ENST00000318560.6 						AGCTGTTATCTGGAAG/AAGCCCTTCAGCGGCC # Human ABL1 across fusion (no fusion)
ENST00000359540.7_BCRABLe1a2					TTCCATGGAGACGCAG/ATGGCTCGTTCGGAAC # Human BCR across fusion at position of e1a2 (no fusion)
ENST00000359540.7_BCRABLe13a2 					ACCATCAATAAGGAAG/ATGATGAGTCTCCGGG # Human BCR across fusion at position of e13a2 (no fusion)
ENST00000359540.7_BCRABLe14a2 					TTTAAGCAGAGTTCAA/ATCTGTACTGCACCCT # Human BCR across fusion at position of e14a2 (no fusion)
BCRABLe1a2 							TTCCATGGAGACGCAG/AAGCCCTTCAGCGGCC # BCR-ABL1 fusion e1a2
BCRABL1e13a2 							ACCATCAATAAGGAAG/AAGCCCTTCAGCGGCC # BCR-ABL1 fusion e13a2
BCRABLe14a2							TTTAAGCAGAGTTCAA/AAGCCCTTCAGCGGCC # BCR-ABL1 fusion e14a2

To allow for deletions at the fusion site we also search for regexes

TTTAAGCAGAGTTCAA/AAGCCCTTCAGCGGCC with up to six deletions or six insertions symmetrical to the junction
                                  that is, up to 12 "free" nucleotides centering 6:6 from the junction viewpoint
 
EOF

echo 'Uncomment to install using conda/similar the environment env/fusion_conda_env.txt'

#conda install micromamba
#mamba init
#micromamba create -n fusion
#micromamba activate fusion
#micromamba install -f env/fusion_conda_env.yaml

# code is to be run on the .bam file generated by the rock_roi_method

## example / simulated run -------------------------------------------------------------------------------------------------

mkdir -p out log

## cannot run the UMI deduplication with UMI tools on the .bam file from STARsolo as we also need the unmapped reads

## run scan on fastq file with cDNA

seqkit locate ./data/fusion_simulations_cdna.fq \
	--use-regexp \
        --pattern-file data/reference_fusions_regex.fa \
        -j $NTHREADS > ./out/fusion_locate_out_reg.txt

## missing aligments (just for simulations)

echo 'Missing alignments'

cut -f2 ./out/fusion_locate_out_reg.txt | grep -v -f - data/fusion_simulations_cdna.fq | grep "@" | wc -l

## from .bam file generate .txt file with read id, cb and ub

## need to do it separate for mapped and unmapped as will have different columns

echo 'Mapped'

echo -e 'Read_id\tCB\tUB' > ./out/reads_with_cb_ub.txt

samtools view -@ $NTHREADS -F 4 ./starsolo/Aligned.sortedByCoord.out.bam | cut -f1,27,28 >> ./out/reads_with_cb_ub.txt

echo 'Unmapped'

samtools view -@ $NTHREADS -f 4 ./starsolo/Aligned.sortedByCoord.out.bam | cut -f1,22,23 >> ./out/reads_with_cb_ub.txt

## append the information to the seqtk file

## need to sort the two files first keeping the header the same

head -1 ./out/fusion_locate_out_reg.txt > ./out/sorted_fusion_locate_out_reg.txt
head -1 ./out/reads_with_cb_ub.txt > ./out/sorted_reads_with_cb_ub.txt
tail -n+2 ./out/fusion_locate_out_reg.txt | sort -k 1,1 >> ./out/sorted_fusion_locate_out_reg.txt
tail -n+2 ./out/reads_with_cb_ub.txt | sort -k 1,1 >> ./out/sorted_reads_with_cb_ub.txt

paste ./out/sorted_fusion_locate_out_reg.txt ./out/sorted_reads_with_cb_ub.txt > ./out/annotated_fusion_locate_out_reg.txt

rm ./out/reads_with_cb_ub.txt ./out/sorted_reads_with_cb_ub.txt ./out/sorted_fusion_locate_out_reg.txt

## UMI deduplication based on start site of alignment

## first removing all alignments with no UB and CB

less ./out/annotated_fusion_locate_out_reg.txt | grep -v 'CB:-' | grep -v 'UB:-' > ./out/cb_ub_annotated_fusion_locate_out_reg.txt

## then sorting based on CB, UMI and pattern name 

sort -k10,10 -k9,9 -k2,2 -u ./out/cb_ub_annotated_fusion_locate_out_reg.txt > ./out/deduplicated_cb_ub_annotated_fusion_locate_out_reg.txt

echo 'Number of patterns prior to deduplication'
wc -l ./out/annotated_fusion_locate_out_reg.txt

echo 'Number of patterns after deduplication'
wc -l ./out/deduplicated_cb_ub_annotated_fusion_locate_out_reg.txt

rm ./out/cb_ub_annotated_fusion_locate_out_reg.txt
