# Experiment

Leukemia patient sample

# Files and folders

Snakefile: used for running rock_roi_method (see https://github.com/imallona/rock_roi_method). The Snakefile was modified to add additional folders in which tmp files are stored and later deleted. 
config.yaml: used for running rock_roi_method (see https://github.com/imallona/rock_roi_method) for the patient sample. 

qmd: files for analysis of count tables and generation of plots
simulation_patient_cell_lines: scripts and simulated .fastq files to test the fusion detection workflow. The simulated reads were generated by hand and were used to test the workflow which was used for the analysis of the datasets from both patient and cell lines. The workflow.sh script includes a STARsolo call to align TSO data without running the full rock_roi_method workflow and the script can thus be run independently. 
workflow: workflow used for analysis of the fusions in the leukemia patient dataset. The fusion detection script is run on the TSO .bam file generated by the rock_roi_method Snakemake workflow and thus needs to be run following this step. 

Files needed to run the simulation_patient_cell_lines workflow:
- simualted .fastq files: the files which are included in the folder contain barcodes for TSO data. Four cells are included - one with minor fusions (and variations), one for major fusions, one for non-fused BCR and ABL and one with reads which contain information which should not be mapped. Both an R1 and an R2 are needed
- custom .fa file: contains the cDNAs covering the fusion region to be detected 
- STAR indexed genome: used to align the simulated data using STARsolo and for running STAR fusion
- genome .gtf: to run STARsolo 
The parameters of interest need to be changed in the in the first lines of the script (# PARAMETERS TO SPECIFY)

Installs neede to run the simulation_patient_cell_lines_workflow:
- STAR: version 2.7.10b was used for the analysis
- bwa: bwa 0.7.18 was used for the analysis

Files needed to run the workflow (on real data): 
- custom .fa file: contains the cDNAs covering the fusion region to be detected. The file is found in the simulation_patient_cell_lines folder. 
- STAR indexed genome: used for running STAR fusion
- TSO .bam file generated with the rock_roi_method
The parameters of interest need to be changed in the in the first lines of the script (# PARAMETERS TO SPECIFY)

Installs neede to run the simulation_patient_cell_lines_workflow:
- STAR: version 2.7.10b was used for the analysis
- bwa: bwa 0.7.18 was used for the analysis


